`"Multiple concurrent writes to Dict detected!"` with `DTables.reduce`

I occasionally get the above error message with the following example. I'm not sure if this issue should go in DTables.jl, but I'm putting it here because the other issues I posted there got migrated here :)

Contents of `mwe.jl`:
```julia
using Distributed
nworkers = 1
addprocs(nworkers - nprocs() + 1)

@everywhere using DTables

remotecall_fetch(2) do
    N = 100
    dt = DTable((a = 1:N, b = rand(N)))
    fetch(reduce(+, dt; cols = [:a]))
end
```

Results:
```julia
julia> include("mwe.jl")
(a = 5050,)

julia> include("mwe.jl")
(a = 5050,)

julia> include("mwe.jl")
(a = 5050,)

julia> include("mwe.jl")
(a = 5050,)

julia> include("mwe.jl")
ERROR: LoadError: On worker 2:
ThunkFailedException:
  Root Exception Type: RemoteException
  Root Exception:
ThunkFailedException:
  Root Exception Type: RemoteException
  Root Exception:
AssertionError: Multiple concurrent writes to Dict detected!
Stacktrace:
  [1] rehash!
    @ ./dict.jl:208
  [2] _setindex!
    @ ./dict.jl:355 [inlined]
  [3] get!
    @ ./dict.jl:477
  [4] reschedule_syncdeps!
    @ ~/.julia/packages/Dagger/ZOt9H/src/sch/util.jl:116
  [5] reschedule_syncdeps!
    @ ~/.julia/packages/Dagger/ZOt9H/src/sch/util.jl:100 [inlined]
  [6] #eager_submit_internal!#96
    @ ~/.julia/packages/Dagger/ZOt9H/src/submission.jl:93
  [7] eager_submit_internal!
    @ ~/.julia/packages/Dagger/ZOt9H/src/submission.jl:11
  [8] #invokelatest#2
    @ ./essentials.jl:819 [inlined]
  [9] invokelatest
    @ ./essentials.jl:816 [inlined]
 [10] #29
    @ ~/.julia/packages/Dagger/ZOt9H/src/sch/dynamic.jl:67 [inlined]
 [11] lock
    @ ./lock.jl:229
 [12] macro expansion
    @ ~/.julia/packages/Dagger/ZOt9H/src/sch/dynamic.jl:66 [inlined]
 [13] #28
    @ ./task.jl:514
Stacktrace:
  [1] exec!
    @ ~/.julia/packages/Dagger/ZOt9H/src/sch/dynamic.jl:108
  [2] eager_submit!
    @ ~/.julia/packages/Dagger/ZOt9H/src/submission.jl:126
  [3] eager_launch!
    @ ~/.julia/packages/Dagger/ZOt9H/src/submission.jl:195
  [4] enqueue!
    @ ~/.julia/packages/Dagger/ZOt9H/src/queue.jl:12 [inlined]
  [5] #spawn#88
    @ ~/.julia/packages/Dagger/ZOt9H/src/thunk.jl:322
  [6] spawn
    @ ~/.julia/packages/Dagger/ZOt9H/src/thunk.jl:286 [inlined]
  [7] #66
    @ ~/.julia/packages/Dagger/ZOt9H/src/thunk.jl:401 [inlined]
  [8] iterate
    @ ./generator.jl:47 [inlined]
  [9] collect
    @ ./array.jl:782
 [10] #65
    @ ~/.julia/packages/DTables/bA4g3/src/operations/operations.jl:146
 [11] #invokelatest#2
    @ ./essentials.jl:819 [inlined]
 [12] invokelatest
    @ ./essentials.jl:816 [inlined]
 [13] #43
    @ ~/.julia/packages/Dagger/ZOt9H/src/processor.jl:162
Stacktrace:
  [1] wait
    @ ./task.jl:349 [inlined]
  [2] fetch
    @ ./task.jl:369 [inlined]
  [3] #execute!#42
    @ ~/.julia/packages/Dagger/ZOt9H/src/processor.jl:172
  [4] execute!
    @ ~/.julia/packages/Dagger/ZOt9H/src/processor.jl:157
  [5] #156
    @ ~/.julia/packages/Dagger/ZOt9H/src/sch/Sch.jl:1551 [inlined]
  [6] #21
    @ ~/.julia/packages/Dagger/ZOt9H/src/options.jl:17 [inlined]
  [7] #1
    @ ~/.julia/packages/ScopedValues/92HJZ/src/ScopedValues.jl:163
  [8] with_logstate
    @ ./logging.jl:514
  [9] with_logger
    @ ./logging.jl:626 [inlined]
 [10] enter_scope
    @ ~/.julia/packages/ScopedValues/92HJZ/src/payloadlogger.jl:17 [inlined]
 [11] with
    @ ~/.julia/packages/ScopedValues/92HJZ/src/ScopedValues.jl:162
 [12] #scoped#4
    @ ./deprecated.jl:116
 [13] scoped
    @ ./deprecated.jl:113
 [14] with_options
    @ ~/.julia/packages/Dagger/ZOt9H/src/options.jl:16
 [15] do_task
    @ ~/.julia/packages/Dagger/ZOt9H/src/sch/Sch.jl:1549
 [16] macro expansion
    @ ~/.julia/packages/Dagger/ZOt9H/src/sch/Sch.jl:1243 [inlined]
 [17] #130
    @ ./task.jl:134
  Root Thunk:  Thunk(id=22, #65(Union{Dagger.EagerThunk, Dagger.Chunk}[Dagger.Chunk{NamedTuple{(:a, :b), Tuple{UnitRange{Int64}, Vector{Float64}}}, MemPool.DRef, Dagger.OSProc, Dagger.AnyScope}(NamedTuple{(:a, :b), Tuple{UnitRange{Int64}, Vector{Float64}}}, Dagger.UnitDomain(), MemPool.DRef(2, 30, 0x0000000000000360), Dagger.OSProc(2), Dagger.AnyScope(), false)], +, [:a], Base._InitialValue()))
  Inner Thunk: Thunk(id=25, #47(+, Thunk[24](#44, Any[:a, Thunk[22](#65, Any[Union{Dagger.EagerThunk, Dagger.Chunk}[Dagger.Chunk{NamedTuple{(:a, :b), Tuple{UnitRange{Int64}, Vector{Float64}}}, MemPool.DRef, Dagger.OSProc, Dagger.AnyScope}(NamedTuple{(:a, :b), Tuple{UnitRange{Int64}, Vector{Float64}}}, Dagger.UnitDomain(), MemPool.DRef(2, 30, 0x0000000000000360), Dagger.OSProc(2), Dagger.AnyScope(), false)], +, [:a], Base._InitialValue()])]), Base._InitialValue()))
  This Thunk:  Thunk(id=25, #47(+, Thunk[24](#44, Any[:a, Thunk[22](#65, Any[Union{Dagger.EagerThunk, Dagger.Chunk}[Dagger.Chunk{NamedTuple{(:a, :b), Tuple{UnitRange{Int64}, Vector{Float64}}}, MemPool.DRef, Dagger.OSProc, Dagger.AnyScope}(NamedTuple{(:a, :b), Tuple{UnitRange{Int64}, Vector{Float64}}}, Dagger.UnitDomain(), MemPool.DRef(2, 30, 0x0000000000000360), Dagger.OSProc(2), Dagger.AnyScope(), false)], +, [:a], Base._InitialValue()])]), Base._InitialValue()))
Stacktrace:
  [1] #fetch#70
    @ ~/.julia/packages/Dagger/ZOt9H/src/eager_thunk.jl:16
  [2] fetch
    @ ~/.julia/packages/Dagger/ZOt9H/src/eager_thunk.jl:11 [inlined]
  [3] #fetch#75
    @ ~/.julia/packages/Dagger/ZOt9H/src/eager_thunk.jl:58 [inlined]
  [4] fetch
    @ ~/.julia/packages/Dagger/ZOt9H/src/eager_thunk.jl:54 [inlined]
  [5] _broadcast_getindex_evalf
    @ ./broadcast.jl:683 [inlined]
  [6] _broadcast_getindex
    @ ./broadcast.jl:656 [inlined]
  [7] getindex
    @ ./broadcast.jl:610 [inlined]
  [8] copy
    @ ./broadcast.jl:912 [inlined]
  [9] materialize
    @ ./broadcast.jl:873 [inlined]
 [10] #50
    @ ~/.julia/packages/DTables/bA4g3/src/operations/operations.jl:115
 [11] #invokelatest#2
    @ ./essentials.jl:819 [inlined]
 [12] invokelatest
    @ ./essentials.jl:816 [inlined]
 [13] #43
    @ ~/.julia/packages/Dagger/ZOt9H/src/processor.jl:162
Stacktrace:
  [1] wait
    @ ./task.jl:349 [inlined]
  [2] fetch
    @ ./task.jl:369 [inlined]
  [3] #execute!#42
    @ ~/.julia/packages/Dagger/ZOt9H/src/processor.jl:172
  [4] execute!
    @ ~/.julia/packages/Dagger/ZOt9H/src/processor.jl:157
  [5] #156
    @ ~/.julia/packages/Dagger/ZOt9H/src/sch/Sch.jl:1551 [inlined]
  [6] #21
    @ ~/.julia/packages/Dagger/ZOt9H/src/options.jl:17 [inlined]
  [7] #1
    @ ~/.julia/packages/ScopedValues/92HJZ/src/ScopedValues.jl:163
  [8] with_logstate
    @ ./logging.jl:514
  [9] with_logger
    @ ./logging.jl:626 [inlined]
 [10] enter_scope
    @ ~/.julia/packages/ScopedValues/92HJZ/src/payloadlogger.jl:17 [inlined]
 [11] with
    @ ~/.julia/packages/ScopedValues/92HJZ/src/ScopedValues.jl:162
 [12] #scoped#4
    @ ./deprecated.jl:116
 [13] scoped
    @ ./deprecated.jl:113
 [14] with_options
    @ ~/.julia/packages/Dagger/ZOt9H/src/options.jl:16
 [15] do_task
    @ ~/.julia/packages/Dagger/ZOt9H/src/sch/Sch.jl:1549
 [16] macro expansion
    @ ~/.julia/packages/Dagger/ZOt9H/src/sch/Sch.jl:1243 [inlined]
 [17] #130
    @ ./task.jl:134
  This Thunk:  Thunk(id=26, #50([:a], Dagger.EagerThunk[EagerThunk (finished)]))
Stacktrace:
  [1] #fetch#70
    @ ~/.julia/packages/Dagger/ZOt9H/src/eager_thunk.jl:16
  [2] fetch
    @ ~/.julia/packages/Dagger/ZOt9H/src/eager_thunk.jl:11 [inlined]
  [3] #fetch#75
    @ ~/.julia/packages/Dagger/ZOt9H/src/eager_thunk.jl:58 [inlined]
  [4] fetch
    @ ~/.julia/packages/Dagger/ZOt9H/src/eager_thunk.jl:54 [inlined]
  [5] #13
    @ ~/tmp/mwe.jl:10
  [6] #invokelatest#2
    @ ./essentials.jl:819 [inlined]
  [7] invokelatest
    @ ./essentials.jl:816
  [8] #110
    @ ~/programs/julia/julia-1.9.3/share/julia/stdlib/v1.9/Distributed/src/process_messages.jl:285
  [9] run_work_thunk
    @ ~/programs/julia/julia-1.9.3/share/julia/stdlib/v1.9/Distributed/src/process_messages.jl:70
 [10] macro expansion
    @ ~/programs/julia/julia-1.9.3/share/julia/stdlib/v1.9/Distributed/src/process_messages.jl:285 [inlined]
 [11] #109
    @ ./task.jl:514
Stacktrace:
 [1] remotecall_fetch(::Function, ::Distributed.Worker; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ Distributed ~/programs/julia/julia-1.9.3/share/julia/stdlib/v1.9/Distributed/src/remotecall.jl:465
 [2] remotecall_fetch(::Function, ::Distributed.Worker)
   @ Distributed ~/programs/julia/julia-1.9.3/share/julia/stdlib/v1.9/Distributed/src/remotecall.jl:454
 [3] #remotecall_fetch#162
   @ ~/programs/julia/julia-1.9.3/share/julia/stdlib/v1.9/Distributed/src/remotecall.jl:492 [inlined]
 [4] remotecall_fetch(::Function, ::Int64)
   @ Distributed ~/programs/julia/julia-1.9.3/share/julia/stdlib/v1.9/Distributed/src/remotecall.jl:492
 [5] top-level scope
   @ ~/tmp/mwe.jl:7
 [6] include(fname::String)
   @ Base.MainInclude ./client.jl:478
 [7] top-level scope
   @ REPL[1]:1
in expression starting at /home/steven/tmp/mwe.jl:7
```

Notice that not every run results in the error.

I also occasionally see the following error printed, but the result of the `remotecall_fetch` still returns normally with the correct answer:
```
Unhandled Task ERROR: ArgumentError: destination has fewer elements than required
Stacktrace:
 [1] copyto!(dest::Vector{Dagger.Sch.ProcessorState}, src::Base.ValueIterator{Dict{Dagger.Processor, Dagger.Sch.ProcessorState}})
   @ Base ./abstractarray.jl:949
 [2] _collect
   @ ./array.jl:713 [inlined]
 [3] collect
   @ ./array.jl:707 [inlined]
 [4] macro expansion
   @ ~/.julia/packages/Dagger/ZOt9H/src/sch/Sch.jl:1189 [inlined]
 [5] (::Dagger.Sch.var"#126#133"{Dagger.Sch.ProcessorInternalState, UInt64, RemoteChannel{Channel{Any}}, Dagger.ThreadProc})()
   @ Dagger.Sch ./task.jl:134
```

This is with DTables v0.4.1 and Dagger v0.18.3.

--- a/src/chunks.jl
+++ b/src/chunks.jl
@@ -65,7 +65,7 @@ affinity(c::Chunk) = affinity(c.handle)
 is_task_or_chunk(c::Chunk) = true
 
 Base.:(==)(c1::Chunk, c2::Chunk) = c1.handle == c2.handle
-Base.hash(c::Chunk, x::UInt64) = hash(c.handle, x)
+Base.hash(c::Chunk, x::UInt64) = hash(c.handle, hash(Chunk, x))
 
 collect_remote(chunk::Chunk) =
     move(chunk.processor, OSProc(), poolget(chunk.handle))
@@ -281,16 +281,22 @@ function savechunk(data, dir, f)
 end
 
 struct WeakChunk
+    wid::Int
+    id::Int
     x::WeakRef
+    function WeakChunk(c::Chunk)
+        return new(c.handle.owner, c.handle.id, WeakRef(c))
+    end
 end
-WeakChunk(c::Chunk) = WeakChunk(WeakRef(c))
 unwrap_weak(c::WeakChunk) = c.x.value
 function unwrap_weak_checked(c::WeakChunk)
-    c = unwrap_weak(c)
-    @assert c !== nothing
-    return c
+    cw = unwrap_weak(c)
+    @assert cw !== nothing "WeakChunk expired: ($(c.wid), $(c.id))"
+    return cw
 end
 is_task_or_chunk(c::WeakChunk) = true
+Serialization.serialize(io::AbstractSerializer, wc::WeakChunk) =
+    error("Cannot serialize a WeakChunk")
 
 Base.@deprecate_binding AbstractPart Union{Chunk, Thunk}
 Base.@deprecate_binding Part Chunk

--- a/src/sch/util.jl
+++ b/src/sch/util.jl
@@ -107,6 +107,18 @@ function reschedule_syncdeps!(state, thunk, seen=Set{Thunk}())
         if haskey(state.cache, thunk) || (thunk in state.ready) || (thunk in state.running)
             continue
         end
+        for (_,input) in thunk.inputs
+            if input isa WeakChunk
+                input = unwrap_weak_checked(input)
+            end
+            if input isa Chunk
+                # N.B. Different Chunks with the same DRef handle will hash to the same slot,
+                # so we just pick an equivalent Chunk as our upstream
+                if !haskey(state.waiting_data, input)
+                    push!(get!(()->Set{Thunk}(), state.waiting_data, input), thunk)
+                end
+            end
+        end
         w = get!(()->Set{Thunk}(), state.waiting, thunk)
         for input in thunk.syncdeps
             input = unwrap_weak_checked(input)

--- a/src/submission.jl
+++ b/src/submission.jl
@@ -27,6 +27,7 @@ function eager_submit_internal!(ctx, state, task, tid, payload; uid_to_tid=Dict{
     timespan_start(ctx, :add_thunk, tid, 0)
 
     # Lookup EagerThunk/ThunkID -> Thunk
+    old_args = copy(args)
     args::Vector{Any}
     syncdeps = if haskey(options, :syncdeps)
         collect(options.syncdeps)
@@ -47,6 +48,13 @@ function eager_submit_internal!(ctx, state, task, tid, payload; uid_to_tid=Dict{
             elseif arg isa Sch.ThunkID
                 arg_tid = arg.id
                 state.thunk_dict[arg_tid]
+            elseif arg isa Chunk
+                # N.B. Different Chunks with the same DRef handle will hash to the same slot,
+                # so we just pick an equivalent Chunk as our upstream
+                if haskey(state.waiting_data, arg)
+                    arg = only(filter(o->o isa Chunk && o.handle == arg.handle, keys(state.waiting_data)))::Chunk
+                end
+                WeakChunk(arg)
             else
                 arg
             end
@@ -76,7 +84,7 @@ function eager_submit_internal!(ctx, state, task, tid, payload; uid_to_tid=Dict{
         options = merge(options, (;syncdeps))
     end
 
-    GC.@preserve args begin
+    GC.@preserve old_args args begin
         # Create the `Thunk`
         thunk = Thunk(f, args...; options...)
 
@@ -146,8 +154,6 @@ function eager_process_elem_submission_to_local(id_map, x)
     @assert !isa(x, Thunk) "Cannot use `Thunk`s in `@spawn`/`spawn`"
     if x isa Dagger.EagerThunk && haskey(id_map, x.uid)
         return Sch.ThunkID(id_map[x.uid], x.thunk_ref)
-    elseif x isa Dagger.Chunk
-        return WeakChunk(x)
     else
         return x
     end

--- a/src/submission.jl
+++ b/src/submission.jl
@@ -1,7 +1,5 @@
 # Remote
 function eager_submit_internal!(@nospecialize(payload))
-    Sch.init_eager()
-
     ctx = Dagger.Sch.eager_context()
     state = Dagger.Sch.EAGER_STATE[]
     task = current_task()
@@ -11,8 +9,6 @@ end
 function eager_submit_internal!(ctx, state, task, tid, payload; uid_to_tid=Dict{UInt64,Int}())
     @nospecialize payload
     ntasks, uid, future, ref, f, args, options, reschedule = payload
-
-    Sch.init_eager()
 
     if uid isa Vector
         thunk_ids = Sch.ThunkID[]
@@ -125,7 +121,14 @@ function eager_submit!(ntasks, uid, future, finalizer_ref, f, args, options)
         h = Dagger.sch_handle()
         return exec!(eager_submit_internal!, h, ntasks, uid, future, finalizer_ref, f, args, options, true)
     elseif myid() != 1
-        return remotecall_fetch(eager_submit_internal!, 1, (ntasks, uid, future, finalizer_ref, f, args, options, true))
+        return remotecall_fetch(1, (ntasks, uid, future, finalizer_ref, f, args, options, true)) do payload
+            @nospecialize payload
+            Sch.init_eager()
+            state = Dagger.Sch.EAGER_STATE[]
+            lock(state.lock) do
+                eager_submit_internal!(payload)
+            end
+        end
     else
         Sch.init_eager()
         state = Dagger.Sch.EAGER_STATE[]

